appvia/kore
# Application service provider

## What

This PR replaces clusterappman with an application service provider and installs the kube-app-manager and the flex-helm-operator components as cluster services in Kore.

## Changes

 * The Service object has now a `Cluster` and `ClusterNamespace` field, which should be filled out if the service lives in a cluster. You can also install something in the `kore` cluster by setting the `Cluster` reference to namespace: `kore-admin` / name: `kore`
 * Clusterappman was completely removed and the manifests were moved to `pkg/kore/assets/application`
 * The application service provider exposes two service kinds:
    * **app** (hidden): used to deploy a list of manifests using template values.
    *. **helm-app**: used to deploy a Helm chart
 * The Application service provider creates two app plans for the two applications
 * The app service kind have Helm-like templating, so simple go template constructs, `{{ .Release.Name }}`, `{{ .Release.Namespace }}` and `{{ .Values.xxx }}` work

## Extensions to the service system

 * a service provider factory can now define default providers, which will be created/updated during setup (this is how we install an application service provider automatically)
 * a service provider can return a list of `Admin services` which should be automatically created. This are basically dependencies of a service provider. Using this we can automatically install `kube-app-manager` and `flex-helm-operator`
 * Admin services are created in the `kore-admin` team, so admins will have visibility over them

## Future use of admin services

 * for the aws-servicebroker provider we can automatically create a Helm Release application, which runs the `aws-servicebroker` app in the Kore cluster
 * for future service operators we can automatically install the operators when a service provider is installed

## Sample application service object

```
➜  kore git:(application_service_provider) ✗ kore get service -t kore-admin -o yaml
apiVersion: services.kore.appvia.io/v1
kind: Service
metadata:
  annotations:
    kore.appvia.io/priority: ""
    kore.appvia.io/serviceProviderName: application
  name: app-flux-helm-operator
  namespace: kore-admin
spec:
  cluster:
    group: clusters.compute.kore.appvia.io
    kind: Cluster
    name: kore
    namespace: kore-admin
    version: v1
  clusterNamespace: flux-helm-operator
  configuration:
    resources: |+
      ---
      apiVersion: app.k8s.io/v1beta1
      kind: Application
      [...]
    values: null
  kind: app
  plan: app-flux-helm-operator
```

## Sample Helm application service object

```
---
apiVersion: services.kore.appvia.io/v1
kind: Service
metadata:
  name: memcached
  namespace: kore-admin
spec:
  kind: helm-app
  summary: Memcached
  description: Memcached Helm Application
  plan: helm-app-memcached
  cluster:
    group: clusters.compute.kore.appvia.io
    kind: Cluster
    name: kore
    namespace: kore-admin
    version: v1
  clusterNamespace: memcached
  configuration:
    source:
      helm:
        url: https://charts.bitnami.com/bitnami
        name: memcached
        version: 4.2.18
    resourceKinds:
      - group: apps
        kind: Deployment
      - group: ""
        kind: Service
```

## How to set up Kore for testing

You will need `kind`.

Create `charts/my_values.yaml`:

```
---
idp:
  client_id: [...]
  client_secret: [...]
  server_url: [...]
api:
  endpoint:
    detect: false
  serviceType: NodePort
  hostPort: 10080
  version: dev
  replicas: 1
  feature_gates: [services=true]
  verbose: true
ui:
  endpoint:
    detect: false
  serviceType: NodePort
  hostPort: 3000
  version: latest
  feature_gates: [services=true]
  replicas: 1
mysql:
  pvc:
    size: 1Gi
```

I've added a new development mode for working with kind, so you have to run the following make script:

```
./scripts/kind_dev.sh --recreate
```

This will delete your existing kind cluster, to make sure you start with a clean slate.

(The other option is to run `USE_GIT_VERSION=true make images push-images`, use the built image version in `my_values.yaml` in `api.version` and you can follow the instructions from `doc/install-kore-in-kind.md`)

## Notes

 * I've found it too late, that we have a new controller component handler for the clusters. We can generalise that and use that for managing service dependencies in the future
 * There is no clean up mechanism for Captain

## Follow up stories

 * Monitor the kube-app-manager (as we can't monitor it with an Application type): https://github.com/appvia/kore/issues/846
 * Document the whole service system: https://github.com/appvia/kore/issues/847
 * ~Deploy Helm charts: https://github.com/appvia/kore/issues/849~
 * UI work for in-cluster and global services: https://github.com/appvia/kore/issues/850
 * Handle secrets: https://github.com/appvia/kore/issues/852

## Testing

The application service provider is installed automatically.

You should check the following:
 * `kore get serviceprovider` lists the `application` provider with status Success
 * `kore get services -t kore-admin` should show the two applications with Success status
 * `kubectl -n kube-app-manager get all` should show components of the app, all as healthy
 * `kubectl -n flu-helm-operator get all` should show components of the app, all as health
 * `kubectl -n flu-helm-operator get application -o yam` should say the application is ready

You can also test removing the service provider which is currently possible:
 * `kore delete serviceprovider application`
 * the admin services should be gone
 * the application namespaces should be empty (we don't delete these)

### Testing a Helm application

1. Create a service plan for Memcached

```
---
apiVersion: services.kore.appvia.io/v1
kind: ServicePlan
metadata:
  name: helm-app-memcached
  namespace: kore
spec:
  kind: helm-app
  summary: Memcached
  description: Memcached Helm Application
  configuration:
    source:
      helm:
        url: https://charts.bitnami.com/bitnami
        name: memcached
        version: 4.2.18
    resourceKinds:
      - group: apps
        kind: Deploymnent
      - group: ""
        kind: Service
```

1. Create a service in `kore-admin`:

```
---
apiVersion: services.kore.appvia.io/v1
kind: Service
metadata:
  name: memcached
  namespace: kore-admin
spec:
  kind: helm-app
  summary: Memcached
  description: Memcached Helm Application
  plan: helm-app-memcached
  cluster:
    group: clusters.compute.kore.appvia.io
    kind: Cluster
    name: kore
    namespace: kore-admin
    version: v1
  clusterNamespace: memcached
  configuration:
    source:
      helm:
        url: https://charts.bitnami.com/bitnami
        name: memcached
        version: 4.2.18
    resourceKinds:
      - group: apps
        kind: Deployment
      - group: ""
        kind: Service
```
